daml-script/test/src/com/digitalasset/daml/lf/engine/script/test/UpgradesIT.scala

```diff
@@ -0,0 +1,365 @@
+        DataDep(
+          path = dar.path,
+          prefix = Some((dar.versionedName, s"V${dar.version}")),
+        )
+      },
+      tmpDir = Some(testCasePkg),
+    ).path
+    (darPath, dars)
+  }
+
+  def assertBuildDar(
+      name: String,
+      version: Int = 1,
+      modules: Map[String, String],
+      deps: Seq[Path] = Seq.empty,
+      dataDeps: Seq[DataDep] = Seq.empty,
+      opts: Seq[String] = Seq(),
+      tmpDir: Option[Path] = None,
+  ): Dar = {
+    val builder = new StringBuilder
+    def log(t: String)(s: String) = discard(builder.append(s"${t}: ${s}\n"))
+    buildDar(
+      name = name,
+      version = version,
+      lfVersion = languageVersion,
+      modules = modules,
+      deps = deps,
+      dataDeps = dataDeps,
+      opts = opts,
+      tmpDir = tmpDir,
+      logger = ProcessLogger(log("stdout"), log("stderr")),
+    ) match {
+      case Right(dar) => dar
+      case Left(exitCode) =>
+        fail(
+          s"While building ${name}-${version}.0.0: 'daml build' exited with ${exitCode}\n${builder.toString}"
+        )
+    }
+  }
+
+  case class TestCase(
+      name: String,
+      damlPath: Path,
+      damlRelPath: Path,
+      pkgDefs: Seq[PackageDefinition],
+  )
+
+  // Ensures no package name is defined twice across all test files
+  def getTestCases(testFileDir: Path): Seq[TestCase] = {
+    val cases = getTestCasesUnsafe(testFileDir)
+    val packageNameDefiners = mutable.Map[String, Seq[String]]()
+    for {
+      c <- cases
+      pkg <- c.pkgDefs
+    } packageNameDefiners.updateWith(pkg.name) {
+      case None => Some(Seq(c.name))
+      case Some(names) => Some(names :+ c.name)
+    }
+    packageNameDefiners.foreach {
+      case (packageName, caseNames) if (caseNames.distinct.length > 1) =>
+        throw new IllegalArgumentException(
+          s"Package with name $packageName is defined multiple times within the following case(s): ${caseNames.distinct
+              .mkString(",")}"
+        )
+      case _ =>
+    }
+    cases
+  }
+
+  def isTest(file: File): Boolean =
+    file.getName.endsWith(".daml")
+
+  def getTestCasesUnsafe(testFileDir: Path): Seq[TestCase] =
+    testFileDir.toFile.listFiles(isTest _).toSeq.map { testFile =>
+      val damlPath = testFile.toPath
+      val damlRelPath = testFileDir.relativize(damlPath)
+      TestCase(
+        name = damlRelPath.toString.stripSuffix(".daml"),
+        damlPath,
+        damlRelPath,
+        pkgDefs = PackageDefinition.readFromFile(damlPath),
+      )
+    }
+
+  case class PackageDefinition(
+      name: String,
+      version: Int,
+      modules: Map[String, String],
+  ) {
+    def build(tmpDir: Path): Dar = {
+      assertBuildDar(
+        name = this.name,
+        version = this.version,
+        modules = this.modules,
+        deps = Seq(damlScriptDar.toPath),
+        tmpDir = Some(tmpDir),
+      )
+    }
+  }
+
+  object PackageDefinition {
```

Comment:
I like this inline extraction logic - it might be nice to ape this for other tests. Can we make a gardening ticket to use this in other parts of the codebase?

Reply:
https://github.com/digital-asset/daml/issues/18579 :)